pkg/resolvers/k8s/watcher.go

```diff
@@ -61,112 +58,132 @@ func (w *watcher) Next() ([]*naming.Update, error) {
+			if ok {
+				return []*naming.Update(nil), errors.Errorf("malformed internal state for endpoints. "+
+					"On added event type, we got update for %v that already exists in %v. Doing resync...", k, w.endpoints)
+			}
+
+			w.endpoints[k] = addr
+			updates = append(updates, &naming.Update{Op: naming.Add, Addr: addr})
 		}
+	case watch.Modified:
+		for k, addr := range endpointsToUpdate {
+			oldAddr, ok := w.endpoints[k]
+			if !ok {
+				return []*naming.Update(nil), errors.Errorf("malformed internal state for endpoints. "+
+					"On modified event type, we got update for %v that does not exists in %v. Doing resync...", k, w.endpoints)
+			}
 
-		updates = append(updates, &naming.Update{Op: naming.Add, Addr: addr, Metadata: md})
-	}
-	// Create updates to delete old endpoints.
-	for addr := range w.lastUpdates {
-		if _, ok := updatedEndpoints[addr]; ok {
-			continue
+			updates = append(updates, &naming.Update{Op: naming.Delete, Addr: oldAddr})
+			w.endpoints[k] = addr
+			updates = append(updates, &naming.Update{Op: naming.Add, Addr: addr})
 		}
-		updates = append(updates, &naming.Update{Op: naming.Delete, Addr: addr, Metadata: nil})
-	}
+	case watch.Deleted:
+		// Create updates to delete old endpoints.
+		for k, addr := range endpointsToUpdate {
+			_, ok := w.endpoints[k]
+			if !ok {
+				return []*naming.Update(nil), errors.Errorf("malformed internal state for endpoints. "+
+					"On delete event type, we got update for %v that does not exists in %v. Doing resync...", k, w.endpoints)
+			}
 
-	w.lastUpdates = updatedEndpoints
+			updates = append(updates, &naming.Update{Op: naming.Delete, Addr: addr})
+			delete(w.endpoints, k)
+		}
+	}
 	return updates, nil
 }
 
-type endpoints struct {
-	Kind       string   `json:"kind"`
-	APIVersion string   `json:"apiVersion"`
-	Metadata   metadata `json:"metadata"`
-	// If kins: Endpoints
-	Subsets []subset `json:"subsets"`
-	// If kind: Status
-	Status  string `json:"status"`
-	Message string `json:"message"`
-	Code    int    `json:"code"`
-}
-
-type metadata struct {
-	Name            string `json:"name"`
-	ResourceVersion string `json:"resourceVersion"`
-}
-
-type subset struct {
-	Addresses []address `json:"addresses"`
-	Ports     []port    `json:"ports"`
-}
-
-type address struct {
-	IP string `json:"ip"`
-}
-
-type port struct {
-	Name string `json:"name"`
-	Port int    `json:"port"`
-}
-
-func subsetToAddresses(t targetEntry, sub subset) ([]string, error) {
+func subsetToAddresses(t targetEntry, sub v1.EndpointSubset) (map[key]string, error) {
 	if len(sub.Ports) == 0 {
-		return []string(nil), errors.Errorf("retrieved subset update contains no port")
+		return nil, errors.Errorf("retrieved subset update contains no port")
 	}
-
 	var port string
+
+	// targetEntry that is specified for this watcher controls what port we should use.
 	if t.port == noTargetPort {
 		// Get first one spotted.
-		port = strconv.Itoa(sub.Ports[0].Port)
+		port = fmt.Sprintf("%v", sub.Ports[0].Port)
```

Comment:
Can we fail if `len(sub.Ports) > 1`? I'm afraid that if we don't do that, it might lead to a hard to fix bugs where kedge usually uses the furst port but seldomly to another different port in a sequence.

Reply:
Hm.. the whole noTargetPort is bit hacky. We should specify target always. So yea, let's restrict the usage, good idea